src/FsAutoComplete/CodeFixes/ChangeDerefBangToValue.fs

```diff
@@ -0,0 +1,82 @@
+/// replace use of ! operator on ref cells with calls to .Value
+module FsAutoComplete.CodeFix.ChangeDerefBangToValue
+
+open FsToolkit.ErrorHandling
+open FsAutoComplete.CodeFix.Types
+open FsAutoComplete
+open FsAutoComplete.LspHelpers
+open FSharp.Compiler.Syntax
+open FSharp.Compiler.Text.Range
+open FSharp.UMX
+
+// let fix (getParseResultsForFile: GetParseResultsForFile) (getLineText: GetLineText): CodeFix =
+//   fun codeActionParams ->
+//     asyncResult {
+//       let fileName = codeActionParams.TextDocument.GetFilePath() |> Utils.normalizePath
+//       let selectionRange = protocolRangeToRange (codeActionParams.TextDocument.GetFilePath()) codeActionParams.Range
+//       let! parseResults, line, lines = getParseResultsForFile fileName selectionRange.Start
+//       let! derefRange = parseResults.GetParseResults.TryRangeOfRefCellDereferenceContainingPos selectionRange.Start |> Result.ofOption (fun _ -> "No deref found at that pos")
+//       let! exprRange = parseResults.GetParseResults.TryRangeOfExpressionBeingDereferencedContainingPos selectionRange.Start |> Result.ofOption (fun _ -> "No expr found at that pos")
+//       let combinedRange = FSharp.Compiler.Text.Range.unionRanges derefRange exprRange
+//       let protocolRange = fcsRangeToLsp combinedRange
+//       let! badString = getLineText lines protocolRange
+//       let replacementString = badString.[1..] + ".Value"
+//       return [
+//         { Title = title
+//           File = codeActionParams.TextDocument
+//           SourceDiagnostic = None
+//           Kind = FixKind.Refactor
+//           Edits = [| { Range = protocolRange
+//                        NewText = replacementString } |] }
+//       ]
+//     }
+
+/// adopted from `dotnet/fsharp` -> `FSharp.Compiler.CodeAnalysis.FSharpParseFileResults.TryRangeOfExpressionBeingDereferencedContainingPos`
+let private tryGetRangeOfDeref input derefPos =
```

Comment:
similarly, this should be replaced with the actual FCS call now that 41.0.3 is released: https://www.fuget.org/packages/FSharp.Compiler.Service/41.0.3/lib/netstandard2.0/FSharp.Compiler.Service.dll/FSharp.Compiler.CodeAnalysis/FSharpParseFileResults?code=true#M%3AFSharp.Compiler.CodeAnalysis.FSharpParseFileResults.TryRangeOfExpressionBeingDereferencedContainingPos%28FSharp.Compiler.Text.Position%29

Reply:
Here it's because trigger/cursor positions are different:
* in `dotnet/fsharp` -> cursor is on expression
* here cursor is on bang

```fsharp
let fr a = ref a
let v = !(fr 5)
//       ^^^^^^ VS and old FSAC trigger range
//      ^ new trigger range
```

-> `dotnet/fsharp` tries to walk down to expression that gets derefed expr, while I walk down to `!`  
-> Check for [`rangeContainsPos expr.Range`](https://github.com/dotnet/fsharp/blob/2f8154178b09f21008e50e32ed58422b54b5789d/src/fsharp/service/FSharpParseFileResults.fs#L339) vs. [`rangeContainsPos funcIdent.idRange` ](https://github.com/Booksbaum/FsAutoComplete/blob/cc2a49f87da2f68b5eb19f17d8d924c3a0044b22/src/FsAutoComplete/CodeFixes/ChangeDerefBangToValue.fs#L40) (-> `!`)

Reason for change: Diagnostics is on bang -> [CodeFix activation via diag](https://github.com/Booksbaum/FsAutoComplete/blob/cc2a49f87da2f68b5eb19f17d8d924c3a0044b22/src/FsAutoComplete/CodeFixes/ChangeDerefBangToValue.fs#L51-L52)  
Otherwise it must always check: is here a `op_Dereference`?

<br/>

(but I forgot to remove some old uncommented code...)

Reply:
ah, I see. That makes sense.